### PR TITLE
Reservation system, testing, etc.

### LINKED ISSUE #67: Vidinė resursų rezervacijos sistema
## Tikslas

Patogus sprendimas, kaip registruoti savo daiktus ir juos skolinti organizacijos viduje.

## Objektai

- [ ] Rezervuojami daiktai
  - Daiktai priklauso padaliniams ir savo daiktus gali redaguoti padalinio administratorius. 
  - Turi būti aprašymo ir nuotraukų laukeliai
  - Daiktai gali būti žymimi, kaip "šiuo metu neskolinami"
  - Objekto struktūra: pavadinimas, vnt. skaičius (iš jo paskaičiuojamas skolintinų vnt. skaičius), daikto fizinė vieta, paveikslėliai, aprašymas
- [ ] Rezervacijos įvykis 
  - Many-to-many su rezervuojamais daiktais ir jų skaičiumi
  - Likusi objekto struktūra: pasiskolinimo data, grąžinimo laikas, aprašymas

## Funkcijos

- [ ] Automatiniai pranešimai rezervacijos įvykio dalyviams (laiškai)
- [ ] Paieška ir filtravimas
- [ ] Būklės valdymas: sukurta -> patvirtinta -> paruošta atsiėmimui -> atsiimta -> grąžinta (approve, reject, cancel)

### Nice to have

- [ ] QR kodai

### PR DESCRIPTION
**Pull request by justinaskav**
This pull request adds a reservation system feature and (finally) tries to use back-end and front-end testing.

Relevent issues:

- Closes #67 

## Reservation system

- Every reservation consists of resources. Resources have their quantity.
- Every reservation and resource must be in Lithuanian, but can be also in English.
- Each resource in a reservations can be approved by resource managers. They are shown for each resource.
- Approval process: Created -> Reserved -> Lent -> Returned
- Resources can be rejected, on Created state. Rejected resources can be deleted for better reservation management.
- Resources can be added to the reservation after it was created.
- Resources may have pictures, which can be added by their managers.
- Resource reservation capacity on specific datetime range is calculated automatically. 
- Everyone can create a reservation, it’s accessible from the admin home page. Every user can be added to a reservation, after it is created. It allows to see reservation comments and add more resources to the reservation.
- Resources can be toggled as not reservable.
- 🅰️ Reservation transitions create tasks (when created, to resource managers, when reserved and lent, to reservation managers)

## Other changes

- Change Mailhog to Mailpit
- Fix Inertia types, better entity types
- Sometimes flash real status code for use in forms
- Optimize decisions in general (affects Doings)
- Optimize ModelPermissionSeeder
- Add better ModelIndexer method, to search on multiple selected fields
- Fix „current duties“ relation method
- Fix meeting index table expand: no more expand all on click
- Add a search button for index datatables
- Add impersonator package (only in local)
- Remove PestPHP, make use of PHPUnit and add some tests (with potential for coverage tests)
    - Because of it, simple migrations were generated instead of schema dump
- Add Vitest and simple ModelChip.test.ts component test case
- Change npm install to npm ci
- Various other styling fixes
- Update all deps (composer, npm, CI/CD)
- 🅰️ Notifications are sent when tasks are created

Todo:

- [x] More resource metadata, in index, edit, reservation table or resource add to reservation
- [x] Check all reservation roles and abilities
- [x] Reservation cards in dashboard
- [x] More clear state flow for users
- [x] Implement Cancel decision for reservation creators
- [x] Check if delete button is implemented correctly (cancel, reject)
- [x] Test for ModelPermissionSeeder, if it doesn’t overwrite existing model permissions
- [x] Decide if leftCapacity method is still needed and implement it better throughout
- [x] More information in reservation index
- [x] Filterable and sortable resources and reservations
- [x] User avatar on form adding user to the reservation
- [x] Try to merge old migrations into one file
    - [x] Test on test process
    - [x] Test on dev setup
- [x] Notification support
- [x] English support

After push to production, the migrations table needs to be emptied and only one record of the first migration needs to be added.

### CODE REVIEW

**Comment by justinaskav** [comment]
@gustaz Dar nėra pabaigta, bet gali apsipažinti jau su platforma

File: app/Exceptions/Handler.php [hunk 1]
```diff
@@ -47,9 +47,16 @@ public function render($request, Throwable $e)
...
+        // TODO: Maybe make errors work something like this: https://inertiajs.com/error-handling
+
+        if (in_array($response->getStatusCode(), [403])) {
+            // check if inertia request
+            if (!$response->headers->get('x-inertia') === 'true') {
+                return $response;
+            }
             return back()->with([
                 'info' => $e->getMessage() ?? 'Neturite teisių atlikti šiam veiksmui.',
+                'statusCode' => $response->getStatusCode(),
```
Comment (mrghosti3) [inline app/Exceptions/Handler.php:59]: Galima perdėlioti metodo logiką iš kažka tokio:

```php
if (!in_array($response->getStatusCode(), [403])
    || !$response->headers->get('x-inertia') === 'true')
{
    return $response;
}

return back()->with([
    'info' => $e->getMessage() ?? 'Neturite teisių atlikti šiam veiksmui.',
    'statusCode' => $response->getStatusCode(),
]);
```

**Comment by justinaskav** [comment]
@gustaz Nieko _major_ nebeplanuoju daryti atm, nebent peržiūrėjus tau ar kitiem norėtųsi dar kažko.

**Comment by mrghosti3** [comment]
Dar trumpas komentaras. Jeigu yra klasės, kurios paveldi (`extend`) kitos klasės `abstract` metodus, bet jų implementacija yra tuščia (does nothing), tai geriau tokia implementaciją dėti ant source. Dirbu prie commit su kuriuom pademonstruosiu ką turiu omenyje.

**Comment by justinaskav** [comment]
@gustaz What is the expectation about tasks? Right now, on reservation resource creation, managers will get the tasks.

**Review by mrghosti3 — CHANGES_REQUESTED** [review]

File: app/Models/Media.php
[the remark this answers is not in the document]
Comment (mrghosti3) [reply app/Models/Media.php:1]: Changes already in #235

**Review by gustaz — APPROVED** [review]

**Comment by gustaz** [comment]
Merged into main

**Comment by mrghosti3** [comment]
Not sure this was merged

**Comment by justinaskav** [comment]
I fix this today

**Pull request closed without merging**